app/src/main/java/org/kiwix/kiwixmobile/zim_manager/local_file_transfer/DeviceListFragment.java

```diff
@@ -0,0 +1,364 @@
+
+    peerDevices.clear();
+    peerDevices.addAll(peers.getDeviceList());
+    ((WifiPeerListAdapter) getListAdapter()).notifyDataSetChanged();
+
+    if (peerDevices.size() == 0) {
+      Log.d(LocalFileTransferActivity.TAG, "No devices found");
+    }
+  }
+
+  public void clearPeers() {
+    peerDevices.clear();
+    ((WifiPeerListAdapter) getListAdapter()).notifyDataSetChanged();
+  }
+
+  void cancelAsyncTasks() {
+    if (peerGroupHandshakeAsyncTask != null) {
+      peerGroupHandshakeAsyncTask.cancel(true);
+    }
+
+    if (senderDeviceAsyncTaskArray != null) {
+      for (SenderDeviceAsyncTask task : senderDeviceAsyncTaskArray) {
+        task.cancel(true);
+      }
+    } else if (receiverDeviceAsyncTask != null) {
+      receiverDeviceAsyncTask.cancel(true);
+    }
+  }
+
+  public void onInitiateDiscovery() { // Setup UI for searching peers
+    searchingPeersProgressBar.setVisibility(View.VISIBLE);
+    listViewPeerDevices.setVisibility(View.INVISIBLE);
+    textViewPeerDevices.setVisibility(View.INVISIBLE);
+  }
+
+  /* From WifiP2pManager.ConnectionInfoListener callback-interface */
+  @Override
+  public void onConnectionInfoAvailable(WifiP2pInfo info) {
+    /* Devices have successfully connected, and 'info' holds information about the wifi p2p group formed */
+    groupInfo = info;
+    // Start handshake between the devices
+    peerGroupHandshakeAsyncTask = new PeerGroupHandshakeAsyncTask(this, groupInfo);
+    peerGroupHandshakeAsyncTask.execute();
+  }
+
+  public void setClientAddress(InetAddress clientAddress) {
+    if (clientAddress == null) {
+      // null is returned only in case of a failed handshake
+      showToast(localFileTransferActivity, R.string.device_not_cooperating, Toast.LENGTH_LONG);
+      localFileTransferActivity.closeLocalFileTransferActivity();
+      return;
+    }
+
+    // If control reaches here, means handshake was successful
+    selectedPeerDeviceInetAddress = clientAddress;
+    startFileTransfer();
+  }
+
+  private void startFileTransfer() {
+    fileTransferStarted = true;
+
+    if (groupInfo.groupFormed && !fileSender) {
+      displayTransferProgressFragment();
+
+      receiverDeviceAsyncTask = new ReceiverDeviceAsyncTask(this, transferProgressFragment);
+      receiverDeviceAsyncTask.execute();
+    } else if (groupInfo.groupFormed) {
+      {
+        Log.d(LocalFileTransferActivity.TAG, "Starting file transfer");
+
+        fileReceiverDeviceAddress =
+            (groupInfo.isGroupOwner) ? selectedPeerDeviceInetAddress : groupInfo.groupOwnerAddress;
+
+        // Hack for allowing slower receiver devices to setup server before sender device requests to connect
+        showToast(localFileTransferActivity, R.string.preparing_files, Toast.LENGTH_LONG);
+        for (int i = 0; i < 20000000; i++) ;
+
+        senderDeviceAsyncTaskArray = new SenderDeviceAsyncTask[totalFilesForTransfer];
+        for (int i = 0; i < totalFilesForTransfer; i++) {
```

Comment:
this is a bit overblown what with creating an array of tasks and all, if you look at `protected Boolean doInBackground(Uri... fileUris)` you will see that AsyncTasks accepts a varargs. It is a bit confusing how that works because you don't invoke that method directly but the hint is in the `AsyncTask.execute` call which also takes a varargs.
example usage would be `new SenderDeviceAsyncTask().execute(fileUriList.toArray())`

If you were hoping that starting all the tasks in a loop would make these operation run in parralel I am afraid you are out of luck, except on Honeycomb but that is a useless piece of information.

Reply:
Thanks for `new SenderDeviceAsyncTask().execute(fileUriList.toArray())`, I'll modify the code

I had a query in:
> If you were hoping that starting all the tasks in a loop would make these operation run in parralel I am afraid you are out of luck, except on Honeycomb but that is a useless piece of information.

Does this mean that _ALL_ async-tasks of a particular type will run sequentially, or only in this case since there is only one socket to communicate with?

Reply:
All asynctasks run sequentially on 1 background thread, this is default behaviour of asynctask. To make it parallel you need to use `executeOnExecutor(Executor)` but I probably wouldn't recommend that.

`fileUriList.toArray()` is psuedocode, it is actually way more annoying in java

Reply:
I remembered why I am using separate asynctasks. It's not for parallelism (that didn't cross my mind until you mentioned it), the reason is Android's inability to provide file sizes from FileProvider uris.

I ran into this problem while learning the P2P API - Basically since the FileProvider uri doesn't correspond to an actual location in the file system, nor  does it provide some "size()" method , I have no way of knowing the file size. There is an InputStream.available() method, but the docs state it has a "weak guarantee" of specifying size correctly. And the naive alternative of creating a new file from the stream will be an overkill to just know the size.

If I don't know the file size, I can't tell the receiver asynctask (one per act of transfer) how many bytes from the sender stream correspond to the first file.
As a result there's no way for the receiver async task to know when the bytes from the first file end and that from the second begin (if I were to use a single sender asynctask).

By using a separate sender asynctask (& hence, a separate connect() request) for each file, the receiver asynctask can be assured that all the bytes from the current sender stream all belong to the same file.

Reply:
@Aditya-Sood I really don't follow. In SenderDeviceAsyncTask you do this:
```
 protected Boolean doInBackground(Uri... fileUris) {
    Uri fileUri = fileUris[0];    // Uri of file to be transferred
```
When I think you could easily do this
```
 protected Boolean doInBackground(Uri... fileUris) {
    for(Uri fileUri : fileUris){
    //your existing code but with a publish call
   }
```

Reply:
[Done](https://github.com/kiwix/kiwix-android/pull/1276/commits/ab8e9da971957063e23c4155b2a7167afac270be#diff-e1d59ae6e15dcb502e125fb6e243d15dR299)